File: src/service-setting-service/src/middleware/tenantIsolation.ts
import { Request, Response, NextFunction } from 'express';
import logger from '../utils/logger';

export const tenantIsolation = (
  req: Request,
  res: Response,
  next: NextFunction
): void => {
  try {
    const user = req.user;
    
    if (!user) {
      res.status(401).json({ error: 'Unauthorized' });
      return;
    }

    const tenantIdFromBody = req.body?.tenantId;
    const tenantIdFromParams = req.params?.tenantId;
    const tenantIdFromQuery = req.query?.tenantId as string;

    const requestedTenantId = tenantIdFromParams || tenantIdFromBody || tenantIdFromQuery;

    if (requestedTenantId && requestedTenantId !== user.tenantId) {
      logger.warn(
        `Tenant isolation violation: User ${user.sub} attempted to access tenant ${requestedTenantId}`
      );
      res.status(403).json({ error: 'Access denied to different tenant' });
      return;
    }

    next();
  } catch (error) {
    logger.error('Tenant isolation error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

File: app/middleware/tenant_isolation.py
from fastapi import HTTPException, status
from app.core.logger import logger
from app.models.auth import JWTPayload
from typing import Optional


def verify_tenant_access(user: JWTPayload, tenant_id: Optional[str]) -> None:
    """Verify user has access to the requested tenant"""
    if tenant_id and tenant_id != user.tenant_id:
        logger.warning(
            f"Tenant isolation violation: User {user.sub} attempted to access tenant {tenant_id}"
        )
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN, detail="Access denied to different tenant"
        )
